sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala

```diff
@@ -88,24 +85,53 @@ case class AnalyzeTableCommand(tableName: String) extends RunnableCommand {
             }
           }.getOrElse(0L)
 
-        // Update the Hive metastore if the total size of the table is different than the size
-        // recorded in the Hive metastore.
-        // This logic is based on org.apache.hadoop.hive.ql.exec.StatsTask.aggregateStats().
-        if (newTotalSize > 0 && newTotalSize != oldTotalSize) {
-          sessionState.catalog.alterTable(
-            catalogTable.copy(
-              properties = relation.catalogTable.properties +
-                (AnalyzeTableCommand.TOTAL_SIZE_FIELD -> newTotalSize.toString)))
-        }
+        updateTableStats(
+          catalogTable,
+          oldTotalSize = catalogTable.stats.map(_.sizeInBytes.toLong).getOrElse(0L),
+          oldRowCount = catalogTable.stats.flatMap(_.rowCount.map(_.toLong)).getOrElse(-1L),
+          newTotalSize = newTotalSize)
+
+      // data source tables have been converted into LogicalRelations
+      case logicalRel: LogicalRelation if logicalRel.catalogTable.isDefined =>
+        updateTableStats(
+          logicalRel.catalogTable.get,
+          oldTotalSize = logicalRel.statistics.sizeInBytes.toLong,
+          oldRowCount = logicalRel.statistics.rowCount.map(_.toLong).getOrElse(-1L),
+          newTotalSize = logicalRel.relation.sizeInBytes)
```

Comment:
Yes, for the first analyze they are equal. But if there's  a second analyze, they are not equal, since logicalRel is in the cachedDataSourceTables, its statistics is the original one.
Anyway, there's a bug here. I think we can use catalogTable's statistics as the old one and I'll restore the test case to check size.